AppActionVisualizer/app/src/main/java/com/example/appactionvisualizer/databean/TestGenerator.java

```diff
@@ -1,85 +1,124 @@
 package com.example.appactionvisualizer.databean;
 
+import android.content.Context;
+import android.util.Log;
+
+import com.example.appactionvisualizer.R;
+import com.example.appactionvisualizer.databean.AppActionProtos.AppAction;
+
+import java.io.IOException;
+import java.io.InputStream;
 import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
 import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
 
 public class TestGenerator {
+  private static final String TAG = "TestGenerator";
+  public static List<AppAction> appActionsUnique = new ArrayList<>();
+  public static Map<ActionType, List<AppAction>> type2appActionList = new HashMap<>();
   private static TestGenerator single_instance = null;
+
   public static TestGenerator getInstance() {
     if (single_instance == null)
       single_instance = new TestGenerator();
     return single_instance;
   }
 
-  /**
-   * used for test the specific app actions
-   * @return dunkin donuts app actions
-   */
-  public AppAction generateDDAppAction() {
-    AppAction newAction = new AppAction();
-    newAction.setAppName("Duckin donuts");
-    newAction.setAppPackage("com.dunkinbrands.otgo");
-    newAction.getActions().add(TestGenerator.getInstance().genTestActionDD1());
-    newAction.getActions().add(TestGenerator.getInstance().genTestActionDD2());
-    newAction.getActions().add(TestGenerator.getInstance().genTestActionDD3());
-    newAction.getActions().add(TestGenerator.getInstance().genTestActionDD4());
-    return newAction;
-  }
+  public void readFromFile(Context context) {
+    InputStream is = context.getResources().openRawResource(R.raw.protobufbinary);
+    try {
+      List<AppAction> appActions = new ArrayList<>();
+      appActions.addAll(AppActionProtos.AppActions.parseFrom(is).getAppActionsList());
+      int sz = appActions.size();
+      int duplicate = 0;
+      for (int i = 0; i < sz; ++i) {
+        boolean isUnique = true;
+        for (int j = i + 1; j < sz; ++j) {
+          if (appActions.get(j).getPackageName().equals(appActions.get(i).getPackageName())) {
+            isUnique = false;
+            duplicate++;
+          }
+        }
+        if (isUnique) {
+          appActionsUnique.add(appActions.get(i));
+        }
+      }
+      int allFulfillmentSize = 0;
+      int zeroParameter = 0, singleParameter = 0, multiParameter = 0;
+      int isCoordinates = 0;
+      Iterator<AppActionProtos.AppAction> it = appActionsUnique.iterator();
```

Comment:
for(AppActionProtos.AppAction appAction : appActionsUnique){}

Reply:
fixed